### ISSUE TITLE
add DPR reader

### ISSUE THREAD

**Issue by MXueguang** [castorini/pygaggle#149]
For now, Pygaggle is for reranking only right?

As we are adding DPR reader, I am thinking about adding a new module for Reader.
So make pygaggle as a mix of reranker and reader.

@KaiSun314 
Basically, what pygaggle doing rn is use a `Reranker` to rerank a list of `Text` for a `Query`. https://github.com/castorini/pygaggle/blob/master/pygaggle/rerank/base.py

To implement the Reader module, we want a `Reader` to read from a list of `Context` for a `Question` and gives an `Answer`, similar to what we did in bertserini (https://github.com/rsvp-ai/bertserini/blob/development/bertserini/reader/base.py)

I guess here, we can use the `Text` class as `Context` and `Query` for `Question`

so concretely, we need to define a data structure class for `Answer` and an abstract class `Reader`. The `DPRReader` should implement the `Reader`. 

When implementing the `DPRReader`, we can use the API from transformers directly, https://huggingface.co/transformers/model_doc/dpr.html#dprreader, which will give us the logits for start/end. 
To get the answer span from the logits, we can learn from how the dpr repo did [here]( https://github.com/facebookresearch/DPR/blob/f403c3b3e179e53c0fe68a0718d5dc25371fe5df/train_reader.py#L304) 

CC: @lintool @ronakice

**Comment by ronakice** [comment]
Yes, and after this add some standard QA metrics like EM/F1 post. We should start with the `TextAnswer` class representing the start/end indices plus the string but ideally, it should also be capable of handling Yes/No type answers (`BooleanAnswer`) and NULL type answers (`NullAnswer`). So this Answer class  will be of type `Union[TextAnswer, BooleanAnswer, NullAnswer]`

**Comment by lintool** [comment]
All sounds good to me at a high level. Let's look at draft PRs to discuss further.

**Comment by KaiSun314** [comment]
On it - reading DPR paper and familiarizing myself with the PyGaggle code, and then will implement a Reader module. ETA: probably a few days.

**Comment by MXueguang** [comment]
closed by https://github.com/castorini/pygaggle/pull/150
